pkg/distributor/ha_tracker.go

```diff
@@ -402,6 +415,34 @@ func (h *haTracker) cleanupOldReplicas(ctx context.Context, deadline time.Time)
 	}
 }
 
+// HaShortcutRequestChecker will check whether the sample in the request must be parsed
+// in the next step or can be shortcut avoiding unneeded request parsing.
+func (h *haTracker) HaShortcutRequestChecker() func(ctx context.Context, w http.ResponseWriter, r *http.Request) bool {
+	return func(ctx context.Context, w http.ResponseWriter, r *http.Request) (finished bool) {
+		cluster, replica := r.Header.Get(ClusterHeader), r.Header.Get(ReplicaHeader)
+		if cluster == "" || replica == "" {
+			return false
+		}
+
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return false
+		}
+
+		// replicaChecker should prefill distributor haTracker.clusters
+		err = h.checkReplica(ctx, userID, cluster, replica, time.Now())
+		if errors.As(err, &replicasNotMatchError{}) {
+			h.haPushRequestSkipParse.Inc()
+			// This sample is coming from secondary replica
+			//// we mark as accepted and send response header
+			w.WriteHeader(http.StatusAccepted)
+			w.Header().Set(IsSecondaryRWReplicaHeader, "true")
```

Comment:
If this was a _real http handler_ we would need to read and discard the entire body, as otherwise we can't use keep-alive.

However, this is httpgrpc, so IDK, but maybe it's worth doing it just to be correct.